src/util/src/integer_util.rs

```diff
@@ -64,6 +66,97 @@ pub fn to_be_bytes_left_pad<T: Borrow<BigUint>>(n: &T, len: usize) -> Vec<u8> {
     v
 }
 
+// Returns the inverse of a_u mod m_u (if it exists)
+pub fn mod_inverse(a_u: &BigUint, m_u: &BigUint) -> Option<BigUint> {
+    if m_u.is_zero() {
+        return None;
+    }
+    let m = BigInt::from_biguint(Sign::Plus, m_u.clone());
+    let mut t = (BigInt::zero(), BigInt::one());
+    let mut r = (m.clone(), BigInt::from_biguint(Sign::Plus, a_u.clone()));
+    while !r.1.is_zero() {
+        let q = r.0.clone() / r.1.clone();
+        //https://docs.rs/num-integer/0.1.45/src/num_integer/lib.rs.html#353
+        let f = |mut r: (BigInt, BigInt)| {
+            mem::swap(&mut r.0, &mut r.1);
+            r.1 -= q.clone() * r.0.clone();
+            r
```

Comment:
I'm a bit surprised that all of these clone's are needed.

Reply:
I have reduced the amount of cloning.